File: src/workshop/utilities.py
import os
from pathlib import Path

from azure.ai.projects.aio import AIProjectClient
from azure.ai.projects.models import MessageImageFileDetails, ThreadMessage, VectorStore


class Utilities:

    def log_msg_green(self, msg: str) -> None:
        print(f"\033[32m{msg}\033[0m")

    def log_msg_purple(self, msg:str) -> None:
        print(f"\033[35m{msg}\033[0m")

    def log_token_blue(self, msg:str) -> None:
        print(f"\033[34m{msg}\033[0m", end="", flush=True)

    async def get_file(self, project_client: AIProjectClient, file_id: str, attachment_name: str) -> None:
        """Retrieve the image file and save it to the local disk."""
        self.log_msg_green(f"Getting file with ID: {file_id}")

        file_name, file_extension = os.path.splitext(os.path.basename(attachment_name.split(":")[-1]))
        file_name = f"{file_name}.{file_id}{file_extension}"

        folder_path = Path("files")
        folder_path.mkdir(exist_ok=True)

        file_path = folder_path / f"{file_name}"

        # Save the file using a synchronous context manager
        with file_path.open("wb") as file:
            async for chunk in await project_client.agents.get_file_content(file_id):
                file.write(chunk)

        self.log_msg_green(f"Image file saved to {file_path}")
        # Cleanup the remote file
        await project_client.agents.delete_file(file_id)

    async def get_files(self, message: ThreadMessage, project_client: AIProjectClient) -> None:
        if message.image_contents:
            for index, image in enumerate(message.image_contents, start=0):
                attachment_name = message.file_path_annotations[index].text
                await self.get_file(project_client, image.image_file.file_id, attachment_name)
        elif message.attachments:
            for index, attachment in enumerate(message.attachments, start=0):
                attachment_name = message.file_path_annotations[index].text
                await self.get_file(project_client, attachment.file_id, attachment_name)


    async def upload_file(self, file_path: str) -> VectorStore:
        """ """
        pass

        # returns a vector_store

        # file = project_client.agents.upload_file_and_poll(file_path="../../datasheet/contoso-tents-datasheet.pdf", purpose="assistants")
        # print(f"Uploaded file, file ID: {file.id}")

        # vector_store = project_client.agents.create_vector_store_and_poll(file_ids=[file.id], name="my_vectorstore")
        # print(f"Created vector store, vector store ID: {vector_store.id}")

File: src/typescript/workshop/src/utilities.ts
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { AIProjectClient } from "@azure/ai-projects";
import { TerminalColors as tc } from './terminalColors.js';

export class Utilities {

  get sharedFilesPath(): string {
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = path.dirname(__filename);
    return path.resolve(__dirname, '../../../shared');
  }

  loadInstructions(instructionsFile: string): string {
    const filePath = path.join(this.sharedFilesPath, instructionsFile);
    return fs.readFileSync(filePath, 'utf8');
  }

  logMsgGreen(msg: string): void {
    console.log(`${tc.GREEN}${msg}${tc.RESET}`);
  }

  logMsgPurple(msg: string): void {
    console.log(`${tc.PURPLE}${msg}${tc.RESET}`);
  }

  logTokenBlue(msg: string): void {
    process.stdout.write(`${tc.BLUE}${msg}${tc.RESET}`);
  }

  private async parseResponse(response: any): Promise<Buffer> {
    if (response instanceof Uint8Array) return Buffer.from(response);
    if (Buffer.isBuffer(response)) return response;
    if (typeof response === 'string') return Buffer.from(response, 'latin1');

    const candidates = ['body', 'data'];
    for (const key of candidates) {
      if (response[key] instanceof Uint8Array) return Buffer.from(response[key]);
      if (typeof response[key] === 'string') return Buffer.from(response[key], 'latin1');
    }

    for (const [key, value] of Object.entries(response)) {
      if (value instanceof Uint8Array) return Buffer.from(value);
      if (typeof value === 'string' && value.length > 0) return Buffer.from(value, 'latin1');
    }

    throw new Error(`Unexpected response format. Keys: ${Object.keys(response).join(', ')}`);
  }

  private ensureFolderExists(folder: string): void {
    if (!fs.existsSync(folder)) {
      fs.mkdirSync(folder, { recursive: true });
    }
  }

  async saveFile(client: AIProjectClient, fileId: string, filePath: string): Promise<void> {
    const response = await client.agents.files.getContent(fileId);
    const fileContent = await this.parseResponse(response);
    await fs.promises.writeFile(filePath, fileContent);
    this.logMsgGreen(`File saved to: ${path.relative(process.cwd(), filePath)}`);
  }

  async downloadFile(client: AIProjectClient, fileId: string, subfolder = 'files'): Promise<void> {
    this.logMsgGreen(`Downloading file with ID: ${fileId}`);
    const filesDir = path.join(this.sharedFilesPath, subfolder);
    this.ensureFolderExists(filesDir);

    let filename = `download.${fileId}.png`;
    try {
      const fileInfo = await client.agents.files.get(fileId);
      if (fileInfo.filename) filename = fileInfo.filename;
    } catch {
      this.logMsgPurple(`Could not get file info, using default filename`);
    }

    const filePath = path.join(filesDir, filename);
    await this.saveFile(client, fileId, filePath);
  }

  async downloadFileWithName(
    client: AIProjectClient,
    fileId: string,
    fileName: string,
    subfolder = 'files'
  ): Promise<void> {
    this.logMsgGreen(`Downloading file with ID: ${fileId} as ${fileName}`);
    const filesDir = path.join(this.sharedFilesPath, subfolder);
    this.ensureFolderExists(filesDir);
    const filePath = path.join(filesDir, fileName);

    const streamableMethod = client.agents.files.getContent(fileId);
    const response = await streamableMethod.asNodeStream();

    if (response.body) {
      const chunks: Buffer[] = [];
      await new Promise<void>((resolve, reject) => {
        response.body!.on('data', (chunk: any) => chunks.push(Buffer.from(chunk)));
        response.body!.on('end', resolve);
        response.body!.on('error', reject);
      });
      await fs.promises.writeFile(filePath, Buffer.concat(chunks));
    } else {
      await this.saveFile(client, fileId, filePath);
    }

    this.logMsgGreen(`File saved to: ${path.relative(process.cwd(), filePath)}`);
  }

  async uploadFile(
    client: AIProjectClient,
    filePath: string,
    purpose = 'assistants'
  ): Promise<any> {
    this.logMsgPurple(`Uploading file: ${filePath}`);
    const fileStream = fs.createReadStream(filePath);
    const fileName = path.basename(filePath);
    const file = await client.agents.files.upload(fileStream, purpose as any, { fileName });
    this.logMsgPurple(`File uploaded with ID: ${file.id}`);
    return file;
  }

  async createVectorStore(
    client: AIProjectClient,
    fileIds: string[],
    vectorStoreName: string
  ): Promise<string> {
    this.logMsgPurple("Creating the vector store");
    const vectorStore = await client.agents.vectorStores.create({ fileIds, name: vectorStoreName });
    this.logMsgPurple(`✅ Created vector store with ID: ${vectorStore.id}`);
    return vectorStore.id;
  }
}
